rasterio/io.py

```diff
@@ -195,6 +202,66 @@ def __repr__(self):
             self.closed and 'closed' or 'open', self.name, self.mode)
 
 
+class MemoryFile(MemoryFileBase):
+    """A BytesIO-like object, backed by an in-memory file.
+
+    This allows formatted files to be read and written without I/O.
+
+    A MemoryFile created with initial bytes becomes immutable. A
+    MemoryFile created without initial bytes may be written to using
+    either file-like or dataset interfaces.
```

Comment:
@brendan-ward I think that's [handled by GDAL](https://github.com/mapbox/rasterio/pull/915/files#diff-49692333e8667a6b046cbdcad3c1164cR846), although I'm not familiar with GDAL's VSI internals.